File: runtimes/java/SovwaveRuntime.java
import java.io.*;
import java.nio.file.*;
import java.util.*;

/**
 * SovwaveRuntime.java — Sovwave MKV Model Runtime (Java 11+)
 *
 * Loads a trained Sovwave .mkv model and runs inference with zero Julia dependency.
 * Requires: Java 11+, ffmpeg in PATH.
 *
 * Usage:
 *   SovwaveModel model = SovwaveModel.load("my_model.mkv", "my_model_meta.yaml");
 *   double[][] outputs = model.predict(new double[][]{{0.1, 0.2, 0.3, 0.4}});
 *   System.out.println(Arrays.toString(outputs[0]));
 *
 * Compile:  javac SovwaveRuntime.java
 * Run:      java SovwaveRuntime my_model.mkv
 */

// ── WaveLayerParams ───────────────────────────────────────────────────────────

class WaveLayerParams {
    int nodes, embedDim;
    double[][] amplitudes;   // [nodes][embedDim]
    double[][] phases;
    double[][] frequencies;
    double[]   fractalScales;
    double[]   fractalDims;   // Hausdorff D_f ∈ [1.0, 3.0]
    double[]   waveSpeeds;    // -1.0 = unlimited
    double omega;

    WaveLayerParams(int nodes, int embedDim, double omega) {
        this.nodes     = nodes;
        this.embedDim  = embedDim;
        this.omega     = omega;
        amplitudes     = new double[nodes][embedDim];
        phases         = new double[nodes][embedDim];
        frequencies    = new double[nodes][embedDim];
        fractalScales  = new double[nodes];
        fractalDims    = new double[nodes];
        waveSpeeds     = new double[nodes];

        for (int i = 0; i < nodes; i++) {
            for (int j = 0; j < embedDim; j++) {
                amplitudes[i][j]  = 0.5;
                phases[i][j]      = 0.0;
                frequencies[i][j] = 1.0;
            }
            fractalScales[i] = 1.618033988749895;
            fractalDims[i]   = 1.5;
            waveSpeeds[i]    = 1.0;
        }
    }
}

// ── SovwaveModel ──────────────────────────────────────────────────────────────

public class SovwaveModel {
    private final List<WaveLayerParams> layers;
    private final int nodes, embedDim, tFrames;
    private final double omega;

    private SovwaveModel(List<WaveLayerParams> layers, int nodes, int embedDim,
                          double omega, int tFrames) {
        this.layers   = layers;
        this.nodes    = nodes;
        this.embedDim = embedDim;
        this.omega    = omega;
        this.tFrames  = tFrames;
    }

    // ── Static loader ─────────────────────────────────────────────────────────

    /**
     * Load a SovwaveModel from an MKV video file.
     * @param mkvPath  Path to the .mkv model file
     * @param metaPath Path to the _meta.yaml config (null = auto-detect)
     */
    public static SovwaveModel load(String mkvPath, String metaPath) throws IOException {
        int    nodes      = 64;
        int    embedDim   = 64;
        int    numLayers  = 3;
        double omega      = 432.0;
        int    tFrames    = 3;
        double betaS      = 1.618033988749895;

        // Auto-detect meta YAML
        String actualMeta = metaPath;
        if (actualMeta == null) {
            actualMeta = mkvPath.replaceAll("(?i)\\.mkv$", "_meta.yaml");
        }
        File metaFile = new File(actualMeta);
        if (metaFile.exists()) {
            try (BufferedReader br = new BufferedReader(new FileReader(metaFile))) {
                String line;
                boolean inModel = false;
                while ((line = br.readLine()) != null) {
                    if (line.trim().equals("model:")) { inModel = true; continue; }
                    if (inModel && !line.isEmpty() && !Character.isWhitespace(line.charAt(0))) inModel = false;
                    if (!inModel) continue;
                    String tl = line.trim();
                    if (tl.startsWith("nodes: "))      nodes      = Integer.parseInt(tl.split(": ")[1]);
                    if (tl.startsWith("embed_dims: ")) embedDim   = Integer.parseInt(tl.split(": ")[1]);
                    if (tl.startsWith("layers: "))     numLayers  = Integer.parseInt(tl.split(": ")[1]);
                    if (tl.startsWith("omega: "))      omega      = Double.parseDouble(tl.split(": ")[1]);
                    if (tl.startsWith("t_frames: "))   tFrames    = Integer.parseInt(tl.split(": ")[1]);
                    if (tl.startsWith("beta_s: "))     betaS      = Double.parseDouble(tl.split(": ")[1]);
                }
            } catch (Exception ignored) {}
        }

        int w = 640;
        int h = 480;

        byte[] raw = extractStream(mkvPath);
        List<WaveLayerParams> layerList = decodePixels(raw, w, h, numLayers, nodes, embedDim, omega, betaS);
        return new SovwaveModel(layerList, nodes, embedDim, omega, tFrames);
    }

    public static SovwaveModel load(String mkvPath) throws IOException {
        return load(mkvPath, null);
    }

    // ── Forward pass ──────────────────────────────────────────────────────────

    private static double[] waveForward(WaveLayerParams layer, double[] input, double t) {
        int n   = layer.nodes;
        int d   = layer.embedDim;
        int ilen = input.length;
        double[] output = new double[n];

        for (int i = 0; i < n; i++) {
            double beta    = layer.fractalScales[i];
            double df      = layer.fractalDims[i];
            double vSpd    = layer.waveSpeeds[i];
            double fracEnv = df / 1.5;
            double nodeSum = 0.0;

            for (int j = 0; j < d; j++) {
                double inVal = j < ilen ? input[j] : 0.5;
                double amp   = layer.amplitudes[i][j];
                double ph    = layer.phases[i][j];
                double freq  = layer.frequencies[i][j];

                double xEff  = (vSpd == -1.0) ? inVal : inVal / Math.max(vSpd, 1e-12);
                double angle = layer.omega * 0.001 * freq * xEff + ph - t;
                nodeSum     += amp * Math.sin(angle);
            }
            output[i] = (nodeSum / Math.sqrt(d)) * beta * fracEnv;
        }
        return output;
    }

    /** Single sample forward pass */
    public double[] forward(double[] input, double t) {
        double[] current = input.clone();
        for (WaveLayerParams layer : layers) {
            double[] accum = new double[layer.nodes];
            for (int frame = 0; frame < tFrames; frame++) {
                double tOff = t + 2 * Math.PI * frame / (omega + 1e-12);
                double[] out = waveForward(layer, current, tOff);
                for (int k = 0; k < layer.nodes; k++) accum[k] += out[k];
            }
            double scale = 1.0 / Math.sqrt(tFrames);
            for (int k = 0; k < accum.length; k++) accum[k] *= scale;
            current = accum;
        }
        return current;
    }

    public double[] forward(double[] input) { return forward(input, 0.0); }

    /** Batch prediction */
    public double[][] predict(double[][] inputs, double t) {
        double[][] results = new double[inputs.length][];
        for (int i = 0; i < inputs.length; i++) results[i] = forward(inputs[i], t);
        return results;
    }

    public double[][] predict(double[][] inputs) { return predict(inputs, 0.0); }

    // ── Pixel decoding ────────────────────────────────────────────────────────

    private static List<WaveLayerParams> decodePixels(
        byte[] raw, int w, int h, int numLayers,
        int nodes, int embedDim, double omega, double betaS
    ) {
        List<WaveLayerParams> layers = new ArrayList<>();
        int bytesPerFrame = w * h * 3;
        double TWO_PI = 2.0 * Math.PI;
        int bw = Math.max(1, w / embedDim);
        int bh = Math.max(1, h / nodes);
        int halfK = Math.max(1, bw / 4);

        for (int l = 0; l < numLayers; l++) {
            WaveLayerParams lp = new WaveLayerParams(nodes, embedDim, omega);
            Arrays.fill(lp.fractalScales, betaS);
            int frameOff = l * bytesPerFrame;

            for (int r = 0; r < nodes; r++) {
                int yc = (int) Math.round((r + 0.5) * bh);
                for (int c = 0; c < embedDim; c++) {
                    int xc = (int) Math.round((c + 0.5) * bw);
                    double rAcc = 0.0, gAcc = 0.0, bAcc = 0.0;
                    int count = 0;
                    for (int dy = -halfK; dy <= halfK; dy++) {
                        for (int dx = -halfK; dx <= halfK; dx++) {
                            int px = Math.min(w - 1, Math.max(0, xc + dx));
                            int py = Math.min(h - 1, Math.max(0, yc + dy));
                            int idx = frameOff + (py * w + px) * 3;
                            if (idx + 2 < raw.length) {
                                rAcc += (raw[idx] & 0xFF);
                                gAcc += (raw[idx + 1] & 0xFF);
                                bAcc += (raw[idx + 2] & 0xFF);
                                count++;
                            }
                        }
                    }
                    if (count > 0) {
                        lp.amplitudes[r][c]  = (rAcc / count / 255.0) * 2.0;
                        lp.phases[r][c]      = (gAcc / count / 255.0) * TWO_PI;
                        lp.frequencies[r][c] = Math.max(0.1, (bAcc / count / 255.0) * 4.0);
                    } else {
                        lp.amplitudes[r][c]  = 0.5;
                        lp.phases[r][c]      = 0.0;
                        lp.frequencies[r][c] = 1.0;
                    }
                }
            }
            layers.add(lp);
        }
        return layers;
    }

    // ── ffmpeg extractor ──────────────────────────────────────────────────────

    private static byte[] extractStream(String mkvPath) throws IOException {
        return runFfmpeg(new String[]{
            "ffmpeg", "-loglevel", "error", "-i", mkvPath,
            "-map", "0:v:0", "-f", "rawvideo", "-pix_fmt", "rgb24", "-"
        });
    }

    private static byte[] runFfmpeg(String[] cmd) throws IOException {
        Process proc = new ProcessBuilder(cmd)
            .redirectErrorStream(false)
            .start();
        byte[] data = proc.getInputStream().readAllBytes();
        try { proc.waitFor(); } catch (InterruptedException e) { Thread.currentThread().interrupt(); }
        return data;
    }

    // ── CLI entrypoint ────────────────────────────────────────────────────────

    public static void main(String[] args) throws Exception {
        if (args.length < 1) {
            System.err.println("Usage: java SovwaveRuntime <model.mkv> [meta.yaml] [in1,in2,...]");
            System.exit(1);
        }
        String mkvPath  = args[0];
        String metaPath = args.length > 1 ? args[1] : null;
        double[] input  = { 0.1, 0.2, 0.3, 0.4 };
        if (args.length > 2) {
            String[] parts = args[2].split(",");
            input = new double[parts.length];
            for (int i = 0; i < parts.length; i++) input[i] = Double.parseDouble(parts[i]);
        }

        SovwaveModel model = SovwaveModel.load(mkvPath, metaPath);
        double[] output = model.forward(input);
        System.out.println("Input:  " + Arrays.toString(input));
        System.out.println("Output: " + Arrays.toString(Arrays.copyOf(output, Math.min(8, output.length))));
    }
}

File: runtimes/csharp/SovwaveRuntime.cs
// SovwaveRuntime.cs — Sovwave MKV Model Runtime (.NET 6+)
//
// Loads a trained Sovwave .mkv model and runs inference with zero Julia dependency.
// Requires: .NET 6+, ffmpeg in PATH.
//
// Usage:
//   var model = await SovwaveModel.LoadAsync("my_model.mkv", "my_model_meta.yaml");
//   double[][] outputs = model.Predict(new[] { new[] { 0.1, 0.2, 0.3, 0.4 } });
//   Console.WriteLine(string.Join(", ", outputs[0]));
//
// Compile: dotnet run  (or add to your .csproj)

using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Sovwave
{
    // ── WaveLayerParams ───────────────────────────────────────────────────────

    internal class WaveLayerParams
    {
        public int Nodes, EmbedDim;
        public double[][] Amplitudes;    // [nodes][embedDim]
        public double[][] Phases;
        public double[][] Frequencies;
        public double[]   FractalScales;
        public double[]   FractalDims;   // Hausdorff D_f ∈ [1.0, 3.0]
        public double[]   WaveSpeeds;    // -1.0 = unlimited
        public double     Omega;

        public WaveLayerParams(int nodes, int embedDim, double omega)
        {
            Nodes        = nodes;
            EmbedDim     = embedDim;
            Omega        = omega;
            Amplitudes   = Jagged(nodes, embedDim, 0.5);
            Phases       = Jagged(nodes, embedDim, 0.0);
            Frequencies  = Jagged(nodes, embedDim, 1.0);
            FractalScales = Enumerable.Repeat(1.618033988749895, nodes).ToArray();
            FractalDims   = Enumerable.Repeat(1.5,               nodes).ToArray();
            WaveSpeeds    = Enumerable.Repeat(1.0,               nodes).ToArray();
        }

        private static double[][] Jagged(int rows, int cols, double fill)
        {
            var m = new double[rows][];
            for (int i = 0; i < rows; i++) {
                m[i] = new double[cols];
                Array.Fill(m[i], fill);
            }
            return m;
        }
    }

    // ── SovwaveModel ──────────────────────────────────────────────────────────

    public class SovwaveModel
    {
        private readonly List<WaveLayerParams> _layers;
        private readonly int    _nodes, _embedDim, _tFrames;
        private readonly double _omega;

        private SovwaveModel(List<WaveLayerParams> layers, int nodes, int embedDim,
                              double omega, int tFrames)
        {
            _layers   = layers;
            _nodes    = nodes;
            _embedDim = embedDim;
            _omega    = omega;
            _tFrames  = tFrames;
        }

        // ── Static loader ─────────────────────────────────────────────────────

        /// <summary>Load a SovwaveModel from an MKV video file.</summary>
        /// <param name="mkvPath">Path to the .mkv model file</param>
        /// <param name="metaPath">Path to the _meta.yaml config (null = auto-detect)</param>
        public static async Task<SovwaveModel> LoadAsync(string mkvPath, string? metaPath = null)
        {
            int    nodes      = 64;
            int    embedDim   = 64;
            int    numLayers  = 3;
            double omega      = 432.0;
            int    tFrames    = 3;
            double betaS      = 1.618033988749895;

            // Auto-detect companion YAML
            string actualMeta = metaPath
                ?? System.Text.RegularExpressions.Regex.Replace(mkvPath, @"(?i)\.mkv$", "_meta.yaml");

            if (File.Exists(actualMeta))
            {
                bool inModel = false;
                foreach (var line in await File.ReadAllLinesAsync(actualMeta))
                {
                    if (line.Trim() == "model:") { inModel = true; continue; }
                    if (inModel && line.Length > 0 && !char.IsWhiteSpace(line[0])) inModel = false;
                    if (!inModel) continue;

                    var tl = line.Trim();
                    if (tl.StartsWith("nodes: "))      nodes      = int.Parse(tl.Split(": ")[1]);
                    if (tl.StartsWith("embed_dims: ")) embedDim   = int.Parse(tl.Split(": ")[1]);
                    if (tl.StartsWith("layers: "))     numLayers  = int.Parse(tl.Split(": ")[1]);
                    if (tl.StartsWith("omega: "))      omega      = double.Parse(tl.Split(": ")[1]);
                    if (tl.StartsWith("t_frames: "))   tFrames    = int.Parse(tl.Split(": ")[1]);
                    if (tl.StartsWith("beta_s: "))     betaS      = double.Parse(tl.Split(": ")[1]);
                }
            }

            int w = 640;
            int h = 480;

            byte[] raw    = await ExtractStreamAsync(mkvPath);
            var layerList = DecodePixels(raw, w, h, numLayers, nodes, embedDim, omega, betaS);
            return new SovwaveModel(layerList, nodes, embedDim, omega, tFrames);
        }

        // ── Forward pass ──────────────────────────────────────────────────────

        private static double[] WaveForward(WaveLayerParams layer, double[] input, double t)
        {
            int n    = layer.Nodes;
            int d    = layer.EmbedDim;
            int ilen = input.Length;
            var output = new double[n];

            for (int i = 0; i < n; i++)
            {
                double beta    = layer.FractalScales[i];
                double df      = layer.FractalDims[i];
                double vSpd    = layer.WaveSpeeds[i];
                double fracEnv = df / 1.5;
                double nodeSum = 0.0;

                for (int j = 0; j < d; j++)
                {
                    double inVal = j < ilen ? input[j] : 0.5;
                    double amp   = layer.Amplitudes[i][j];
                    double ph    = layer.Phases[i][j];
                    double freq  = layer.Frequencies[i][j];

                    // Speed-aware phase: -1.0 = unlimited
                    double xEff  = vSpd == -1.0 ? inVal : inVal / Math.Max(vSpd, 1e-12);
                    double angle = layer.Omega * 0.001 * freq * xEff + ph - t;
                    nodeSum     += amp * Math.Sin(angle);
                }
                output[i] = (nodeSum / Math.Sqrt(d)) * beta * fracEnv;
            }
            return output;
        }

        /// <summary>Single-sample forward pass through all layers.</summary>
        public double[] Forward(double[] input, double t = 0.0)
        {
            double[] current = (double[])input.Clone();
            foreach (var layer in _layers)
            {
                var accum = new double[layer.Nodes];
                for (int frame = 0; frame < _tFrames; frame++)
                {
                    double tOff = t + 2.0 * Math.PI * frame / (_omega + 1e-12);
                    var    @out = WaveForward(layer, current, tOff);
                    for (int k = 0; k < layer.Nodes; k++) accum[k] += @out[k];
                }
                double scale = 1.0 / Math.Sqrt(_tFrames);
                for (int k = 0; k < accum.Length; k++) accum[k] *= scale;
                current = accum;
            }
            return current;
        }

        /// <summary>Batch prediction.</summary>
        public double[][] Predict(double[][] inputs, double t = 0.0)
            => inputs.Select(inp => Forward(inp, t)).ToArray();

        // ── Pixel decoding ────────────────────────────────────────────────────

        private static List<WaveLayerParams> DecodePixels(
            byte[] raw, int w, int h, int numLayers,
            int nodes, int embedDim, double omega, double betaS)
        {
            var layers        = new List<WaveLayerParams>();
            int bytesPerFrame = w * h * 3;
            double twoPi      = 2.0 * Math.PI;
            int bw            = Math.Max(1, w / embedDim);
            int bh            = Math.Max(1, h / nodes);
            int halfK         = Math.Max(1, bw / 4);

            for (int l = 0; l < numLayers; l++)
            {
                var lp = new WaveLayerParams(nodes, embedDim, omega);
                Array.Fill(lp.FractalScales, betaS);
                int frameOff = l * bytesPerFrame;

                for (int r = 0; r < nodes; r++)
                {
                    int yc = (int)Math.Round((r + 0.5) * bh);
                    for (int c = 0; c < embedDim; c++)
                    {
                        int xc = (int)Math.Round((c + 0.5) * bw);
                        double rAcc = 0.0, gAcc = 0.0, bAcc = 0.0;
                        int count = 0;
                        for (int dy = -halfK; dy <= halfK; dy++)
                        {
                            for (int dx = -halfK; dx <= halfK; dx++)
                            {
                                int px = Math.Min(w - 1, Math.Max(0, xc + dx));
                                int py = Math.Min(h - 1, Math.Max(0, yc + dy));
                                int idx = frameOff + (py * w + px) * 3;
                                if (idx + 2 < raw.Length)
                                {
                                    rAcc += raw[idx];
                                    gAcc += raw[idx + 1];
                                    bAcc += raw[idx + 2];
                                    count++;
                                }
                            }
                        }
                        if (count > 0)
                        {
                            lp.Amplitudes[r][c]  = (rAcc / count / 255.0) * 2.0;
                            lp.Phases[r][c]      = (gAcc / count / 255.0) * twoPi;
                            lp.Frequencies[r][c] = Math.Max(0.1, (bAcc / count / 255.0) * 4.0);
                        }
                        else
                        {
                            lp.Amplitudes[r][c]  = 0.5;
                            lp.Phases[r][c]      = 0.0;
                            lp.Frequencies[r][c] = 1.0;
                        }
                    }
                }
                layers.Add(lp);
            }
            return layers;
        }

        // ── ffmpeg extractor ──────────────────────────────────────────────────

        private static async Task<byte[]> ExtractStreamAsync(string mkvPath)
        {
            return await RunFfmpegAsync(
                "ffmpeg",
                $"-loglevel error -i \"{mkvPath}\" -map 0:v:0 -f rawvideo -pix_fmt rgb24 -");
        }

        private static async Task<byte[]> RunFfmpegAsync(string exe, string args)
        {
            var psi = new ProcessStartInfo(exe, args)
            {
                RedirectStandardOutput = true,
                RedirectStandardError  = true,
                UseShellExecute        = false,
                CreateNoWindow         = true
            };
            using var proc = Process.Start(psi) ?? throw new Exception("Failed to start ffmpeg");
            using var ms   = new MemoryStream();
            await proc.StandardOutput.BaseStream.CopyToAsync(ms);
            await proc.WaitForExitAsync();
            return ms.ToArray();
        }

        // ── CLI entrypoint ────────────────────────────────────────────────────

        public static async Task Main(string[] args)
        {
            if (args.Length < 1) {
                Console.Error.WriteLine("Usage: SovwaveRuntime <model.mkv> [meta.yaml] [0.1,0.2,...]");
                return;
            }
            string mkvPath  = args[0];
            string? meta    = args.Length > 1 ? args[1] : null;
            double[] input  = args.Length > 2
                ? args[2].Split(',').Select(double.Parse).ToArray()
                : new[] { 0.1, 0.2, 0.3, 0.4 };

            var    model  = await SovwaveModel.LoadAsync(mkvPath, meta);
            double[] output = model.Forward(input);
            Console.WriteLine($"Input:  [{string.Join(", ", input)}]");
            Console.WriteLine($"Output: [{string.Join(", ", output.Take(8).Select(v => $"{v:F6}"))}]");
        }
    }
}
